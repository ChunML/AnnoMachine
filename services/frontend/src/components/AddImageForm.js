import React from 'react';
import PropTypes from 'prop-types';

class AddImageForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      image_url: '',
      image_file: '',
    };

    this.fileInput = React.createRef();

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleFileUpload = this.handleFileUpload.bind(this);
    this.handleButtonClick = this.handleButtonClick.bind(this);
  }

  handleInputChange(e) {
    const { name, value } = e.target;
    this.setState({
      [name]: value,
    });
  }

  handleFileUpload(e) {
    const file = e.target.files[0];
    this.setState({
      image_file: file,
    });
  }

  handleButtonClick(e) {
    e.preventDefault();
    const { onButtonClick } = this.props;
    const data = { ...this.state };
    if (this.fileInput.current) {
      this.fileInput.current.value = '';
    }
    this.setState({
      image_url: '',
      image_file: null,
    });
    onButtonClick(data);
  }

  render() {
    const { image_url } = this.state;
    return (
      <form className="form" encType="multipart/form-data">
        <div className="input-field">
          <input
            name="image_file"
            type="file"
            onChange={this.handleFileUpload}
            ref={this.fileInput}
          />
        </div>
        <div className="input-field">
          <input
            name="image_url"
            type="text"
            placeholder="Or enter an image URL"
            value={image_url}
            onChange={this.handleInputChange}
          />
        </div>
        <button
          className="primary button"
          type="submit"
          onClick={this.handleButtonClick}
        >
          Submit
        </button>
      </form>
    );
  }
}

AddImageForm.propTypes = {
  onButtonClick: PropTypes.func.isRequired,
};

export default AddImageForm;
