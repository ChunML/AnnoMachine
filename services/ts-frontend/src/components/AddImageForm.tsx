import React, { useState } from "react";
import PropTypes from 'prop-types';

interface AddImageFormProps {
  onButtonClick(data: AddImageFormStates): void;
}

interface AddImageFormStates {
  image_url: string;
  image_file: null | File;
}

function AddImageForm(props: AddImageFormProps) {
  // fileInput: React.RefObject<HTMLInputElement>;
  const fileInput = React.createRef<HTMLInputElement>();
  const [image_url, setImageUrl] = useState("");
  const [image_file, setImageFile] = useState<File | null>(null);

  const handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    setImageUrl(e.currentTarget.value);
  };

  const handleFileUpload = (e: React.FormEvent<HTMLInputElement>) => {
    if (!e.currentTarget.files?.length) return;
    const file = e.currentTarget.files[0];
    setImageFile(file);
  };

  const handleButtonClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    const { onButtonClick } = props;
    const data = { image_url, image_file };
    setImageUrl("");
    setImageFile(null);
    if (fileInput.current) {
      fileInput.current.value = "";
    }
    onButtonClick(data);
  };

  return (
    <form className="form" encType="multipart/form-data">
      <div className="input-field">
        <input
          name="image_file"
          type="file"
          onChange={handleFileUpload}
          ref={fileInput}
        />
      </div>
      <div className="input-field">
        <input
          name="image_url"
          type="text"
          placeholder="Or enter an image URL"
          value={image_url}
          onChange={handleInputChange}
        />
      </div>
      <button
        className="primary button"
        type="submit"
        onClick={handleButtonClick}
      >
        Submit
      </button>
    </form>
  );
}

AddImageForm.propTypes = {
  onButtonClick: PropTypes.func.isRequired,
};


export default AddImageForm;
