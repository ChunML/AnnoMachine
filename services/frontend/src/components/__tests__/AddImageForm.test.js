import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import AddImageForm from '../AddImageForm';
import '../../setupTests';

const onButtonClick = jest.fn();

it('AddImageForm renders properly', () => {
  const wrapper = shallow(<AddImageForm onButtonClick={onButtonClick} />);
  const form = wrapper.find('form');
  expect(form.length).toBe(1);
  const fields = form.find('.input-field');
  expect(fields.length).toBe(2);
  expect(fields.get(0).props.children.type).toBe('input');
  expect(fields.get(0).props.children.props.name).toBe('image_file');
  expect(fields.get(1).props.children.type).toBe('input');
  expect(fields.get(1).props.children.props.name).toBe('image_url');
  expect(onButtonClick).toHaveBeenCalledTimes(0);

  const image_url = form.find('input[name="image_url"]');
  image_url.simulate('change', { target: { name: 'image_url', value: 'abc' } });
  const button = form.find('button');
  button.simulate('click', { preventDefault: () => {} });
  expect(onButtonClick).toHaveBeenCalledTimes(1);

  const image_file = form.find('input[name="image_file"]');
  image_file.simulate('change', { target: { files: ['file_1'] } });
  image_url.simulate('change', { target: { name: 'image_url', value: '' } });
  button.simulate('click', { preventDefault: () => {} });
  expect(onButtonClick).toHaveBeenCalledTimes(2);
});

it('AddImageForm renders a snapshot properly', () => {
  const tree = renderer
    .create(
      <Router location="/">
        <AddImageForm onButtonClick={jest.fn()} />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
