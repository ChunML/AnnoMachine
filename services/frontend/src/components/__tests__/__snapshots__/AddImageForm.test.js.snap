// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddImageForm renders a snapshot properly 1`] = `
<form
  className="ui form"
  encType="multipart/form-data"
>
  <div
    className="field"
  >
    <input
      name="image_file"
      onChange={[Function]}
      type="file"
    />
  </div>
  <div
    className="field"
  >
    <input
      name="image_url"
      onChange={[Function]}
      placeholder="Enter an image URL"
      type="text"
      value=""
    />
  </div>
  <div
    className="field"
  >
    <div
      className="ui checkbox"
    >
      <input
        name="is_private"
        type="checkbox"
      />
      <label>
        Private
      </label>
    </div>
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</form>
`;
